chain/client/src/sync/header.rs

```diff
@@ -237,75 +234,51 @@ impl HeaderSync {
         None
     }
 
+    // The remote side will return MAX_BLOCK_HEADERS headers, starting from the first hash in
+    // the returned "locator" list that is on their canonical chain.
+    //
+    // The locator allows us to start syncing from a reasonably recent common ancestor. Since
+    // we don't know which fork the remote side is on, we include a few hashes. The first one
+    // we include is the tip of our chain, and the next one is 2 blocks back, then 4 blocks
+    // back, then 8 blocks back, etc, until we reach the most recent final block. The reason
+    // why we stop at the final block is because the consensus guarantees us that the final
+    // blocks observed by all nodes are on the same fork.
     fn get_locator(&mut self, chain: &Chain) -> Result<Vec<CryptoHash>, near_chain::Error> {
-        let tip = chain.header_head()?;
-        let genesis_height = chain.genesis().height();
-        let heights = get_locator_heights(tip.height - genesis_height)
-            .into_iter()
-            .map(|h| h + genesis_height)
-            .collect::<Vec<_>>();
-
-        // For each height we need, we either check if something is close enough from last locator, or go to the db.
-        let mut locator: Vec<(u64, CryptoHash)> = vec![(tip.height, tip.last_block_hash)];
-        for h in heights {
-            if let Some(x) = close_enough(&self.history_locator, h) {
-                locator.push(x);
-            } else {
-                // Walk backwards to find last known hash.
-                let last_loc = *locator.last().unwrap();
-                if let Ok(header) = chain.get_block_header_by_height(h) {
-                    if header.height() != last_loc.0 {
-                        locator.push((header.height(), *header.hash()));
-                    }
-                }
-            }
+        let store = chain.store();
+        let tip = store.header_head()?;
+        // We could just get the ordinal from the header, but it's off by one: #8177.
+        let tip_ordinal = store.get_block_merkle_tree(&tip.last_block_hash)?.size();
+        let final_head = store.final_head()?;
+        let final_head_ordinal = store.get_block_merkle_tree(&final_head.last_block_hash)?.size();
```

Comment:
Added a new test. It turns out this is pretty tricky to do because constructing a correct block is tricky. Fortunately the earlier test in this file was doing this properly (but in fact it had a mistake and the test didn't need the correct setup anyway - which I simplified down). The new test requires 5 syncs.